stored_requests/fetcher.go

```diff
@@ -19,52 +19,98 @@ type Fetcher interface {
 
 // Cache is an intermediate layer which can be used to create more complex Fetchers by composition.
 // Implementations must be safe for concurrent access by multiple goroutines.
-// To add a Cache layer in front of the Fetcher, see WithCache()
+// To add a Cache layer in front of a Fetcher, see WithCache()
 type Cache interface {
-	// GetRequests works much like Fetcher.FetchRequests, with a few exceptions:
+	// Get works much like Fetcher.FetchRequests, with a few exceptions:
 	//
 	// 1. Any errors should be logged by the implementation, rather than returned.
 	// 2. The returned map _may_ be written to.
 	// 3. The returned map must _not_ contain keys unless they were present in the argument ID list.
 	// 4. Callers _should not_ assume that the returned map contains a key for every id.
 	//    The returned map will miss entries for keys which don't exist in the cache.
-	GetRequests(ctx context.Context, ids []string) map[string]json.RawMessage
+	Get(ctx context.Context, ids []string) map[string]json.RawMessage
 
-	// SaveRequests stores some data in the cache. The map is from ID to the cached value.
-	//
-	// This is a best-effort method. If the cache call fails, implementations should log the error.
-	SaveRequests(ctx context.Context, values map[string]json.RawMessage)
+	// Invalidate will ensure that all values associated with the given IDs
+	// are no longer returned by the cache until new values are saved via Update
+	Invalidate(ctx context.Context, ids []string)
+
+	// Update will update the given values in the cache
+	Update(ctx context.Context, data map[string]json.RawMessage)
+}
+
+// ComposedCache creates an interface to treat a slice of caches as a single cache
+type ComposedCache []Cache
+
+// Get will attempt to Get from the caches in the order in which they are in the slice,
+// stopping as soon as a value is found (or when all caches have been exhausted)
+func (c ComposedCache) Get(ctx context.Context, ids []string) (data map[string]json.RawMessage) {
+	data = make(map[string]json.RawMessage)
+	remainingIds := ids
+
+	for _, cache := range c {
+		if cachedData := cache.Get(ctx, remainingIds); len(cachedData) > 0 {
+			// iterate over remainingIds from end, droppings ids as they are filled
+			for i := len(remainingIds) - 1; i >= 0; i-- {
+				if config, ok := cachedData[remainingIds[i]]; ok {
+					data[remainingIds[i]] = config
+					remainingIds = append(remainingIds[:i], remainingIds[i+1:]...)
+				}
+			}
+		}
+
+		// return if all ids filled
+		if len(remainingIds) == 0 {
+			return
+		}
+	}
+	return
+}
+
+// Invalidate will propagate invalidations to all underlying caches
+func (c ComposedCache) Invalidate(ctx context.Context, ids []string) {
+	for _, cache := range c {
+		cache.Invalidate(ctx, ids)
+	}
+}
+
+// Update will propagate updates to all underlying caches
+func (c ComposedCache) Update(ctx context.Context, data map[string]json.RawMessage) {
```

Comment:
If update will propagate to all underlying caches, who then need to decide if it is a valid update or not (i.e. check to see if the key exists first), then we need a separate `Save()` method to tell the cache "This is a new value you are supposed to save". Otherwise the cache will either save every update that comes through, no matter if it has ever been requested or not, or never save anything as all updates fail the "is already in cache" check.

Reply:
It's a lot of overhead to check determine the union of keys in the update and in the cache at each layer and then only forward the subset of updates to the cache's `Update`. Two alternatives that come to mind:
1. Leave as-is and let individual cache implementations decide what to do with an `Update` (e.g. LRUCache's Update could just ignore updates for keys that aren't present, or aggressively cache if it's relatively empty and ignore updates for non-existent keys when relatively full)
2. Change the signature of `Update` to just pass in a single `string` and `json.RawMessage` instead of a batch `map[string]json.RawMessage`.